File: generate_event_tracking_data.py
import uuid
import random
from datetime import datetime, timedelta
from faker import Faker
from faker_commerce import Provider as CommerceProvider
from random import randint

# Initialize Faker with a locale
fake = Faker('en_US')

# Add the commerce provider to the faker instance
fake.add_provider(CommerceProvider)

# Generate user_info Table Data
def generate_user_data():
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    profile = fake.profile()
    age = fake.random_int(min=18, max=90) 
    user_id = f"user_id-{uuid.uuid4()}"
    password_default = fake.password()
    user_info_data = {
        "id": user_id,
        "username": profile['username'],
        "real_name": profile['name'],
        "phone_number": fake.phone_number(),
        "sex": profile['sex'],
        "job": profile['job'],
        "company": profile['company'],
        "email": profile['mail'],
        "password":password_default,
        "birth_of_date": profile['birthdate'],
        "age": age,
        "create_time": create_time,
        "delete_time": delete_time
    }
    return user_info_data

# Generate Address Table Data
def generate_fake_address(user):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    address_titles = [
        "Home Address",
        "Work Address",
        "Billing Address",
        "Shipping Address",
        "Vacation Home"
    ]
    street_address = fake.street_address()
    secondary_address = fake.secondary_address()
    full_address_line = f"{street_address} {secondary_address}"
    address_id = f"address_id-{uuid.uuid4()}"
    address_data = {
        "id": address_id,
        "user_id":user["id"],
        "title": random.choice(address_titles),
        "address_line": full_address_line,
        "country": fake.country(),
        "city": fake.city(),
        "postal_code": fake.postcode(),
        "create_time": create_time,
        "delete_time": delete_time
    }
    
    return address_data


# Generate Categories Table Data
def generate_categories_data():
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    category_id = f"category_id-{uuid.uuid4()}"
    category_name = fake.ecommerce_category()
    description = fake.text(max_nb_chars=10)
    categories_data = {
        "id": category_id, 
        "name": category_name,
        "description": description,
        "create_time": create_time,
        "delete_time": delete_time
    }
    return categories_data

# Generate Subcategories Table Data
def generate_subcategories_data(category):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    subcategory_id = f"subcategory_id-{uuid.uuid4()}"
    subcategory_name = fake.unique.word().capitalize() + " Subcategory"
    description = fake.text(max_nb_chars=50)
    subcategories_data = {
        "id": subcategory_id,
        "parent_id": category["id"],
        "name": subcategory_name,
        "description": description,
        "create_time": create_time,
        "delete_time": delete_time
    }
    return subcategories_data

# Generate Products Table Data
def generate_products_data(subcategory):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    product_id = f"product_id-{uuid.uuid4()}"
    product_name = fake.unique.catch_phrase()
    product_description = fake.paragraph()  
    products_data = {
        "id": product_id,
        "name": product_name,
        "description": product_description,
        "category_id": subcategory["id"],  
        "create_time": create_time,
        "delete_time": delete_time
    }
    return products_data

# Generate product_sku Table Data
def generate_sku_data(category,subcategory,product):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    sku_id = f"sku_id-{uuid.uuid4()}"
    sku_number = fake.unique.random_number(digits=5, fix_len=True)
    sku = f"{category['id'][-3:]}-{subcategory['id'][-3:]}-{product['id'][-3:]}-{sku_id[-3:]}"
    price = round(random.uniform(5.0, 500.0), 2)  
    quantity = random.randint(0, 1000)          
    
    skus_data = {
        "id": sku_id,
        "product_id": product["id"],
        "sku": sku,
        "price": price,
        "quantity": quantity,
        "create_time": create_time,
        "delete_time": delete_time
        }

    return skus_data

# Generate Wishlist Table Data
def generate_wishlist_data(product,user):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    delete_time = fake.date_time_between(start_date=create_time, end_date='now')
    wishlist_id = f"wishlist_id-{uuid.uuid4()}"
    wishlists_data = {
        'id': wishlist_id,
        'user_id': user["id"],
        'product_id': product["id"],
        "create_time": create_time,
        "delete_time": delete_time
            }
    return wishlists_data

# Generate Cart Table Data
def generate_cart_data(user,products_sku,cartItem):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    cart_id = f"cart_id-{uuid.uuid4()}"
    total_price = cartItem["quantity"] * products_sku["price"]
    carts_data = {
        'id': cart_id,
        'user_id': user["id"],
        'total':total_price,
        "create_time": create_time,
        'updated_at': fake.date_time_between(start_date='-6m', end_date='now')
        }
    return carts_data

# Generate CartItem Table Data
def generate_cartItem_data(cart,product,products_sku):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    cartItem_id = f"cartItem_id-{uuid.uuid4()}"
    cart_items_data = {
        'id': cartItem_id,
        'cart_id': cart["id"],
        'product_id': product["id"],
        'products_sku_id': products_sku["id"],
        'quantity': random.randint(1, 5),
        "create_time": create_time,
        'updated_at': fake.date_time_between(start_date='-6m', end_date='now')
        }
    return cart_items_data

# Generate Order_details Table Data
def generate_Order_details_data(order_item,user,products_sku):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    order_details_id = f"order_details_id-{uuid.uuid4()}"
    order_details_data = {
        'id': order_details_id,
        'user_id': user["id"],
        'payment_id':0,
        'total': order_item["quantity"] * products_sku["price"],
        "create_time": create_time,
        'updated_at': fake.date_time_between(start_date='-6m', end_date='now')
        }
    return order_details_data


# Generate Order_item Table Data
def generate_Order_item_data(product,products_sku,order):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    order_item_id = f"order_item_id-{uuid.uuid4()}"
    order_item_data = {
        'id': order_item_id,
        'order_id': order["id"],
        'product_id': product["id"],
        'products_sku_id': products_sku["id"],
        'quantity': random.randint(1, 99999999),
        "create_time": create_time,
        'updated_at': fake.date_time_between(start_date='-6m', end_date='now')
        }
    return order_item_data


# Generate Payment_details Table Data
def generate_Payment_details_data(order_details):
    create_time = fake.date_time_between(start_date='-2y', end_date='now')
    payment_details_id = f"payment_details_id-{uuid.uuid4()}"
    payment_statuses = ['Success', 'Pending', 'Failed', 'Refunded']
    if order_details[''] is None:
        payment_amount = 0
    else:
        payment_amount = order_details['total']
    payment_details_data = {
        'id': payment_details_id,
        'order_details_id': order_details["id"],
        'amount': payment_amount,
        'provider':fake.credit_card_full().split('\n')[0].strip(),
        'status':random.choice(payment_statuses) ,
        "create_time": create_time,
        'updated_at': fake.date_time_between(start_date='-6m', end_date='now')
        }
    return payment_details_data  




    
print('sss')

File: backend/backend .NET/GenerateEventTrackingData.cs
using System;
using System.Collections.Generic;
using Bogus;

namespace MyProject.Backend.DataGeneration
{
    // POCO classes representing the generated entities
    public record UserInfo(string Id, string Username, string RealName, string PhoneNumber, string Sex, string Job, string Company, string Email, string Password, DateTime BirthOfDate, int Age, DateTime CreateTime, DateTime DeleteTime);

    public record Address(string Id, string UserId, string Title, string AddressLine, string Country, string City, string PostalCode, DateTime CreateTime, DateTime DeleteTime);

    public record Category(string Id, string Name, string Description, DateTime CreateTime, DateTime DeleteTime);

    public record Subcategory(string Id, string ParentId, string Name, string Description, DateTime CreateTime, DateTime DeleteTime);

    public record Product(string Id, string Name, string Description, string CategoryId, DateTime CreateTime, DateTime DeleteTime);

    public record Sku(string Id, string ProductId, decimal Price, long Quantity, DateTime CreateTime, DateTime DeleteTime);

    public record Wishlist(string Id, string UserId, string ProductsSkuId, DateTime CreateTime, DateTime DeleteTime);

    public record Cart(string Id, string OrderId, string ProductsSkuId, int Quantity, DateTime CreateTime, DateTime UpdatedAt);

    public record OrderDetails(string Id, string UserId, string PaymentId, DateTime CreateTime, DateTime UpdatedAt);

    public record OrderItem(string Id, string OrderId, string ProductsSkuId, long Quantity, DateTime CreateTime, DateTime UpdatedAt);

    public record PaymentDetails(string Id, decimal Amount, string Provider, string Status, DateTime CreateTime, DateTime UpdatedAt);


    // DataGenerator - uses Bogus to mimic logic from the Python generator
    // Usage:
    // 1. Create a .NET project (console/library) and add the Bogus NuGet package:
    //    dotnet add package Bogus
    // 2. Include this file in the project and call the DataGenerator methods.
    public class DataGenerator
    {
        private readonly Faker _faker;
        private readonly Random _random;

        public DataGenerator()
        {
            _faker = new Faker("en_US");
            _random = new Random();
        }

        public UserInfo GenerateUserData()
        {
            var createTime = DateTime.Now;
            // pick a delete_time between createTime and now (may equal createTime)
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var birthOfDate = _faker.Date.PastOffset(_faker.Random.Int(18, 75) + 0, DateTime.Now).DateTime.Date;
            var age = DateTime.Now.Year - birthOfDate.Year;
            var userId = $"user_id-{Guid.NewGuid()}";
            var passwordDefault = _faker.Internet.Password();

            return new UserInfo(
                Id: userId,
                Username: _faker.Internet.UserName(),
                RealName: _faker.Name.FullName(),
                PhoneNumber: _faker.Phone.PhoneNumber(),
                Sex: _faker.PickRandom(new[] { "M", "F" , "Other" }),
                Job: _faker.Name.JobTitle(),
                Company: _faker.Company.CompanyName(),
                Email: _faker.Internet.Email(),
                Password: passwordDefault,
                BirthOfDate: birthOfDate,
                Age: age,
                CreateTime: createTime,
                DeleteTime: deleteTime
            );
        }

        public Address GenerateFakeAddress(UserInfo user)
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var titles = new[] { "Home Address", "Work Address", "Billing Address", "Shipping Address", "Vacation Home" };
            var fullAddressLine = $"{_faker.Address.StreetAddress()} {_faker.Address.SecondaryAddress()}";
            var addressId = $"address_id-{Guid.NewGuid()}";

            return new Address(
                Id: addressId,
                UserId: user.Id,
                Title: _faker.PickRandom(titles),
                AddressLine: fullAddressLine,
                Country: _faker.Address.Country(),
                City: _faker.Address.City(),
                PostalCode: _faker.Address.ZipCode(),
                CreateTime: createTime,
                DeleteTime: deleteTime
            );
        }

        public Category GenerateCategoriesData()
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var categoryId = $"category_id-{Guid.NewGuid()}";
            var categoryName = _faker.Commerce.Categories(1)[0];
            var description = _faker.Lorem.Sentence(10);

            return new Category(categoryId, categoryName, description, createTime, deleteTime);
        }

        public Subcategory GenerateSubcategoriesData(Category category)
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var subcategoryId = $"subcategory_id-{Guid.NewGuid()}";
            // Replace .Capitalize() with char.ToUpper(word[0]) + word.Substring(1)
            var word = _faker.Lorem.Word();
            var subcategoryName = char.ToUpper(word[0]) + word.Substring(1) + " Subcategory";
            var description = _faker.Lorem.Sentence(8);

            return new Subcategory(subcategoryId, category.Id, subcategoryName, description, createTime, deleteTime);
        }

        public Product GenerateProductsData(Subcategory subcategory)
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var productId = $"product_id-{Guid.NewGuid()}";
            var productName = _faker.Company.CatchPhrase();
            var productDescription = _faker.Lorem.Paragraph();

            return new Product(productId, productName, productDescription, subcategory.Id, createTime, deleteTime);
        }

        public Sku GenerateSkuData(Category category, Subcategory subcategory, Product product)
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var skuNumber = _faker.Random.Number(10000, 99999);
            var skuId = $"{category.Id[^3..]}-{subcategory.Id[^3..]}-{product.Id[^3..]}-{skuNumber}";
            var price = Math.Round((decimal)(_faker.Random.Decimal(5.0m, 500.0m)), 2);
            var stock = _faker.Random.Long(0, 9_999_999);

            return new Sku(skuId, product.Id, price, stock, createTime, deleteTime);
        }

        public Wishlist GenerateWishlistData(Sku productsSku, UserInfo user)
        {
            var createTime = DateTime.Now;
            var deleteTime = _faker.Date.Between(createTime, DateTime.Now);
            var wishlistId = $"wishlist_id-{Guid.NewGuid()}";

            return new Wishlist(wishlistId, user.Id, productsSku.Id, createTime, deleteTime);
        }

        public Cart GenerateCartData(Sku productsSku, OrderDetails order)
        {
            var createTime = DateTime.Now;
            var updateTime = _faker.Date.Between(createTime, DateTime.Now);
            var cartId = $"cart_id-{Guid.NewGuid()}";
            var quantity = _faker.Random.Int(1, 9999);

            return new Cart(cartId, order.Id, productsSku.Id, quantity, createTime, updateTime);
        }

        public OrderDetails GenerateOrderDetailsData(UserInfo user, PaymentDetails payment)
        {
            var createTime = DateTime.Now;
            var updateTime = _faker.Date.Between(createTime, DateTime.Now);
            var orderDetailsId = $"order_details_id-{Guid.NewGuid()}";

            return new OrderDetails(orderDetailsId, user.Id, payment.Id, createTime, updateTime);
        }

        public OrderItem GenerateOrderItemData(Sku productsSku, OrderDetails order)
        {
            var createTime = DateTime.Now;
            var updateTime = _faker.Date.Between(createTime, DateTime.Now);
            var orderItemId = $"order_item_id-{Guid.NewGuid()}";
            var quantity = _faker.Random.Long(1, 99_999_999);

            return new OrderItem(orderItemId, order.Id, productsSku.Id, quantity, createTime, updateTime);
        }

        public PaymentDetails GeneratePaymentDetailsData()
        {
            var createTime = DateTime.Now;
            var updateTime = _faker.Date.Between(createTime, DateTime.Now);
            var paymentDetailsId = $"payment_details_id-{Guid.NewGuid()}";
            var paymentStatuses = new[] { "Success", "Pending", "Failed", "Refunded" };

            // provider: use credit card provider/brand as a short provider name
            var provider = _faker.Finance.CreditCardNumber();

            return new PaymentDetails(paymentDetailsId, 0m, provider, _faker.PickRandom(paymentStatuses), createTime, updateTime);
        }
    }
}
